plugins/amdgpu/amdgpu_plugin.c

```diff
@@ -528,6 +529,116 @@ void amdgpu_plugin_fini(int stage, int ret)
+	pid_t pid;
+	struct kfd_criu_bo_buckets *bo_buckets;
+	BoEntriesTest **bo_info_test;
+	int ret;
+};
+
+void *dump_bo_contents(void *_thread_data)
+{
+        int i, ret = 0;
+        int num_bos = 0;
+        struct thread_data* thread_data = (struct thread_data*) _thread_data;
+        struct kfd_criu_bo_buckets *bo_buckets = thread_data->bo_buckets;
+        BoEntriesTest **bo_info_test = thread_data->bo_info_test;
+
+        pr_info("amdgpu_plugin: Thread[0x%x] started\n", thread_data->gpu_id);
+        for (i = 0; i < thread_data->num_of_bos; i++) {
+		if (bo_buckets[i].gpu_id != thread_data->gpu_id)
+                       continue;
+
+		num_bos++;
+		if (bo_buckets[i].bo_alloc_flags & KFD_IOC_ALLOC_MEM_FLAGS_VRAM ||
+		    bo_buckets[i].bo_alloc_flags & KFD_IOC_ALLOC_MEM_FLAGS_GTT) {
+			char *fname;
+			int mem_fd;
+
+			if (bo_info_test[i]->bo_alloc_flags & KFD_IOC_ALLOC_MEM_FLAGS_PUBLIC) {
+				int drm_fd;
+				void *addr;
+				struct tp_node *dev;
+
+				plugin_log_msg("amdgpu_plugin: large bar read possible\n");
+
+				dev = sys_get_node_by_gpu_id(&src_topology, bo_buckets[i].gpu_id);
+				if (!dev) {
+					ret = -EFAULT;
+					goto failed;
+				}
+
+				drm_fd = open_drm_render_device(dev->drm_render_minor);
+				if (drm_fd < 0) {
+					ret = -EFAULT;
+					goto failed;
+				}
+
+				addr = mmap(NULL,
+						bo_buckets[i].bo_size,
+						PROT_READ,
+						MAP_SHARED,
+						drm_fd,     /* mapping on local gpu for prototype */
+						bo_buckets[i].bo_offset);
+				if (addr == MAP_FAILED) {
+					pr_perror("amdgpu_plugin: mmap failed\n");
+					close(drm_fd);
+					goto failed;
+				}
+
+				/* direct memcpy is possible on large bars */
+				memcpy(bo_info_test[i]->bo_rawdata.data,
+					addr, bo_buckets[i].bo_size);
+				munmap(addr, bo_buckets[i].bo_size);
+				close(drm_fd);
+			} else {
+				plugin_log_msg("Now try reading BO contents with /proc/pid/mem");
+				if (asprintf (&fname, PROCPIDMEM, thread_data->pid) < 0) {
+					pr_perror("failed in asprintf, %s\n", fname);
+					ret = -1;
+					goto failed;
+				}
+
+				mem_fd = open (fname, O_RDONLY);
+				if (mem_fd < 0) {
+					pr_perror("Can't open %s for pid %d\n", fname, thread_data->pid);
+					free (fname);
+					ret = -1;
+					goto failed;
+				}
+
+				plugin_log_msg("Opened %s file for pid = %d\n", fname, thread_data->pid);
+				free (fname);
+				if (lseek (mem_fd, (off_t) bo_buckets[i].bo_addr, SEEK_SET) == -1) {
+					pr_perror("Can't lseek for bo_offset for pid = %d\n", thread_data->pid);
+					ret = -1;
+					goto failed;
+				}
+				plugin_log_msg("Try to read file now\n");
+
+				if (read(mem_fd, bo_info_test[i]->bo_rawdata.data,
+					bo_info_test[i]->bo_size) != bo_info_test[i]->bo_size) {
+					pr_perror("Can't read buffer\n");
+					ret = -1;
+					goto failed;
+				}
+
+				close(mem_fd);
+			} /* PROCPIDMEM read done */
+               }
+        }
+	pr_info("amdgpu_plugin: Thread[0x%x] done num_bos:%d ret:%d\n", thread_data->gpu_id, num_bos, ret);
```

Comment:
Indentation is off. It looks like your editor is set to tab-width 4 and getting mixed up when to indent with tabs or spaces. For kernel code (and I think CRIU uses the same convention), TAB width should be 8 and we should always use TABs for indentation where possible.

Reply:
ACK